File: Java/oraclestorage/src/main/java/com/ithit/webdav/samples/oraclestorageservlet/websocket/WebSocketServer.java
package com.ithit.webdav.samples.oraclestorageservlet.websocket;

import com.ithit.webdav.samples.oraclestorageservlet.WebDavEngine;
import com.ithit.webdav.server.util.StringUtil;

import javax.servlet.http.HttpSession;
import javax.websocket.*;
import javax.websocket.server.ServerEndpoint;
import java.util.HashSet;
import java.util.Set;

/**
 * WebSocket server, creates web socket endpoint, handles client's sessions
 */
@ServerEndpoint(value = "/",
        configurator = GetHttpSessionConfigurator.class, encoders = {NotificationEncoder.class})
public class WebSocketServer {

    private static final Set<Session> SESSIONS = new HashSet<>();
    private HttpSession httpSession;

    @OnOpen
    public void onOpen(Session session, EndpointConfig config) {
        SESSIONS.add(session);
        httpSession = (HttpSession) config.getUserProperties()
                .get(HttpSession.class.getName());
        ((WebDavEngine) httpSession.getAttribute("engine")).setWebSocketServer(this);
    }

    @OnMessage
    public String onMessage(String message) {
        return message;
    }

    @OnClose
    public void onClose(Session session) {
        SESSIONS.remove(session);
        ((WebDavEngine) httpSession.getAttribute("engine")).setWebSocketServer(this);
    }

    /**
     * Send notification to the client
     *
     * @param itemPath   File/Folder path.
     * @param operation  Operation name: created/updated/deleted/moved
     */
    private void send(String itemPath, String operation) {
        itemPath = StringUtil.trimEnd(StringUtil.trimStart(itemPath, "/"), "/");
        final Notification notification = new Notification(itemPath, operation);
        for (Session s : SESSIONS) {
            if (s.isOpen()) {
                s.getAsyncRemote().sendObject(notification);
            }
        }
    }

    /**
     * Notifies client that file/folder was created.
     *
     * @param itemPath file/folder.
     */
    public void notifyCreated(String itemPath) {
        send(itemPath, "created");
    }

    /**
     * Notifies client that file/folder was updated.
     *
     * @param itemPath file/folder.
     */
    public void notifyUpdated(String itemPath) {
        send(itemPath, "updated");
    }

    /**
     * Notifies client that file/folder was deleted.
     *
     * @param itemPath file/folder.
     */
    public void notifyDeleted(String itemPath) {
        send(itemPath, "deleted");
    }

    /**
     * Notifies client that file/folder was locked.
     *
     * @param itemPath file/folder.
     */
    public void notifyLocked(String itemPath) {
        send(itemPath, "locked");
    }

    /**
     * Notifies client that file/folder was unlocked.
     *
     * @param itemPath file/folder.
     */
    public void notifyUnlocked(String itemPath) {
        send(itemPath, "unlocked");
    }

    /**
     * Notifies client that file/folder was moved.
     *
     * @param itemPath file/folder.
     */
    public void notifyMoved(String itemPath, String targetPath) {
        itemPath = StringUtil.trimEnd(StringUtil.trimStart(itemPath, "/"), "/");
        targetPath = StringUtil.trimEnd(StringUtil.trimStart(targetPath, "/"), "/");
        final MovedNotification movedNotification = new MovedNotification(itemPath, "moved", targetPath);
        for (Session s : SESSIONS) {
            if (s.isOpen()) {
                s.getAsyncRemote().sendObject(movedNotification);
            }
        }
    }

    /**
     * Represents VO to exchange between client and server
     */
    static class Notification {
        private final String itemPath;
        private final String operation;

        Notification(String itemPath, String operation) {
            this.itemPath = itemPath;
            this.operation = operation;
        }

        String getItemPath() {
            return itemPath;
        }

        String getOperation() {
            return operation;
        }
    }

    /**
     * Represents VO to exchange between client and server for move type
     */
    static class MovedNotification extends Notification {
        private final String targetPath;

        MovedNotification(String itemPath, String operation, String targetPath) {
            super(itemPath, operation);
            this.targetPath = targetPath;
        }

        String getTargetPath() {
            return targetPath;
        }

    }
}

File: Kotlin/filesystemstorage/src/main/kotlin/com/ithit/webdav/samples/fsstorageservlet/websocket/WebSocketServer.kt
package com.ithit.webdav.samples.fsstorageservlet.websocket

import com.ithit.webdav.samples.fsstorageservlet.WebDavEngine
import com.ithit.webdav.server.util.StringUtil
import java.util.*
import javax.servlet.http.HttpSession
import javax.websocket.*
import javax.websocket.server.ServerEndpoint

/**
 * WebSocket server, creates web socket endpoint, handles client's sessions
 */
@ServerEndpoint(value = "/", configurator = GetHttpSessionConfigurator::class, encoders = [NotificationEncoder::class])
class WebSocketServer {
    private var httpSession: HttpSession? = null

    @OnOpen
    fun onOpen(session: Session, config: EndpointConfig) {
        sessions.add(session)
        httpSession = config.userProperties[HttpSession::class.java.name] as HttpSession
        (httpSession!!.getAttribute("engine") as WebDavEngine).webSocketServer = this
    }

    @OnMessage
    fun onMessage(message: String): String {
        return message
    }

    @OnClose
    fun onClose(session: Session) {
        sessions.remove(session)
        (httpSession!!.getAttribute("engine") as WebDavEngine).webSocketServer = this
    }

    /**
     * Send notification to the client
     *
     * @param itemPath   File/Folder path.
     * @param operation  Operation name: created/updated/deleted/moved
     */
    private fun send(itemPath: String, operation: String) {
        var itemPath: String? = itemPath
        itemPath = StringUtil.trimEnd(StringUtil.trimStart(itemPath, "/"), "/")
        var notification = Notification(itemPath, operation)
        for (s in sessions) {
            if (s.isOpen) {
                s.asyncRemote.sendObject(notification)
            }
        }
    }

    /**
     * Notifies client that file/folder was created.
     *
     * @param itemPath file/folder.
     */
    fun notifyCreated(itemPath: String) {
        send(itemPath, "created")
    }

    /**
     * Notifies client that file/folder was updated.
     *
     * @param itemPath file/folder.
     */
    fun notifyUpdated(itemPath: String) {
        send(itemPath, "updated")
    }

    /**
     * Notifies client that file/folder was deleted.
     *
     * @param itemPath file/folder.
     */
    fun notifyDeleted(itemPath: String) {
        send(itemPath, "deleted")
    }

    /**
     * Notifies client that file/folder was locked.
     *
     * @param itemPath file/folder.
     */
    fun notifyLocked(itemPath: String) {
        send(itemPath, "locked")
    }

    /**
     * Notifies client that file/folder was unlocked.
     *
     * @param itemPath file/folder.
     */
    fun notifyUnlocked(itemPath: String) {
        send(itemPath, "unlocked")
    }

    /**
     * Notifies client that file/folder was moved.
     *
     * @param itemPath file/folder.
     */
    fun notifyMoved(itemPath: String?, targetPath: String?) {
        var itemPath = itemPath
        var targetPath = targetPath
        itemPath = StringUtil.trimEnd(StringUtil.trimStart(itemPath, "/"), "/")
        targetPath = StringUtil.trimEnd(StringUtil.trimStart(targetPath, "/"), "/")
        var movedNotification = MovedNotification(itemPath, "moved", targetPath)
        for (s in sessions) {
            if (s.isOpen) {
                s.asyncRemote.sendObject(movedNotification)
            }
        }
    }

    /**
     * Represents VO to exchange between client and server
     */
    open inner class Notification(val itemPath: String?, val operation: String)

    /**
     * Represents VO to exchange between client and server for move type
     */
    inner class MovedNotification(itemPath: String?, operation: String, val targetPath: String?) :
        Notification(itemPath, operation)

    companion object {
        private val sessions: MutableSet<Session> = HashSet()
    }
}
